src/lib/utils.ts

```diff
@@ -106,6 +106,43 @@ export function deferExecution<T>(func: (args: T) => any, wait: number) {
   return deferred
 }
 
+/**
+ * Replace a function's return type with another type.
+ */
+type WithReturnType<F extends (...args: any[]) => any, NewReturn> = (
+  ...args: Parameters<F>
+) => NewReturn
+
+/**
+ * Assert that a function type is async, preserving its parameter types.
+ */
+type AsyncFn<F extends (...args: any[]) => any> = WithReturnType<
+  F,
+  Promise<unknown>
+>
```

Comment:
These might live better in `lib/types.ts` but I'm not sure best practice since these are only used here. They do seem general enough to be at-risk for being duplicated if needed elsewhere, because someone doesn't know about these local definitions.